README.md
AWS CloudFront URL Signature Utility
===================
[![Build Status](https://travis-ci.org/jasonsims/aws-cloudfront-sign.svg?branch=master)](https://travis-ci.org/jasonsims/aws-cloudfront-sign)
[![npm version](https://badge.fury.io/js/aws-cloudfront-sign.svg)](http://badge.fury.io/js/aws-cloudfront-sign)

---

**NOTE**

The [AWS SDK for JavaScript](https://docs.aws.amazon.com/en_pv/sdk-for-javascript/v2/developer-guide/welcome.html) has added support for generating signed URLs and Cookies.  Please see [Class: AWS.CloudFront.Signer](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFront/Signer.html)

---

Generating signed URLs for CloudFront links is a little more tricky than for S3. It's because signature generation for S3 URLs is handled a bit differently than CloudFront URLs and this functionality is not currently supported by the [aws-sdk](https://github.com/aws/aws-sdk-js) library for JavaScript. In case you also need to do this, I've created this simple utility to make things easier.

## Usage
### Requirements
* Node.js >=18
* Active CloudFront distribution with origin configured

### Configuring CloudFront
1. Create a CloudFront distribution
2. Configure your origin with the following settings:

   **Origin Domain Name:** {your-s3-bucket}
   **Restrict Bucket Access:** Yes
   **Grant Read Permissions on Bucket:** Yes, Update Bucket Policy
3. Create CloudFront Key Pair. [more info][cf_keypair_docs]

### Installing
```sh
npm install aws-cloudfront-sign
```

### TypeScript
```js
import { SignatureOptions } from 'aws-cloudfront-sign/types'
```

### Upgrading from 2.x to 3.x
* There shouldn't be any breaking changes when coming to 3.x. RTMP URLs were deprecated by Amazon
but that will affect all versions.
* Support for ES Modules was added
* Support for TypeScript was added

### Upgrading from 1.x to 2.x
* `expireTime` now takes it's value as milliseconds, Date, or
 [moment][moment_docs] instead of seconds.

### API
#### getSignedUrl(url, options)
* `@param {String} url` - Cloudfront URL to sign
* `@param {Object} options` - URL signature [options](#options)
* `@return {String} signedUrl` - Signed CloudFrontUrl

#### getSignedCookies(url, options)
* `@param {String} url` - Cloudfront URL to sign
* `@param {Object} options` - URL signature [options](#options)
* `@return {Object} cookies` - Signed AWS cookies

#### ~~getSignedRTMPUrl(domainName, s3key, options)~~
⛔️ **Deprecated**: [RTMP Support Discontinuing on December 31, 2020](https://repost.aws/questions/QUoUZgHZh7SEWlnQUPlBmVNQ/announcement-rtmp-support-discontinuing-on-december-31-2020)
* ~~`@param {String} domainName` - Domain name of your Cloudfront distribution~~
* ~~`@param {String} s3key` - Path to s3 object~~
* ~~`@param {Object} options` - URL signature [options](#options)~~
* ~~`@return {Object} url.rtmpServerPath` - RTMP formatted server path~~
* ~~`@return {Object} url.rtmpStreamName` - Signed RTMP formatted stream name~~

### Options
* `expireTime` (**Optional** - Default: 1800 sec == 30 min) - The time when the URL should expire. Accepted values are
   * number - Time in milliseconds (`new Date().getTime() + 1800000`)
   * moment - Valid [momentjs][moment_docs] object (`moment().add(1, 'day')`)
   * Date - Javascript Date object (`new Date(2016, 0, 1)`)
* `ipRange` (**Optional**) - IP address range allowed to make GET requests
  for your signed URL. This value must be given in standard IPv4 CIDR format
  (for example, 10.52.176.0/24).
* `keypairId` - The access key ID from your Cloudfront keypair
* `privateKeyString` || `privateKeyPath` - The private key from your Cloudfront
   keypair. It can be provided as either a string or a path to the .pem file.
  **Note:** When providing the private key as a string, ensure that the newline
  character is also included.

  ```js
  const privateKeyString =
    '-----BEGIN RSA PRIVATE KEY-----\n'
    'MIIJKAIBAAKCAgEAwGPMqEvxPYQIffDimM9t3A7Z4aBFAUvLiITzmHRc4UPwryJp\n'
    'EVi3C0sQQKBHlq2IOwrmqNiAk31/uh4FnrRR1mtQm4x4IID58cFAhKkKI/09+j1h\n'
    'tuf/gLRcOgAXH9o3J5zWjs/y8eWTKtdWv6hWRxuuVwugciNckxwZVV0KewO02wJz\n'
    'jBfDw9B5ghxKP95t7/B2AgRUMj+r47zErFwo3OKW0egDUpV+eoNSBylXPXXYKvsL\n'
    'AlznRi9xNafFGy9tmh70pwlGG5mVHswD/96eUSuLOZ2srcNvd1UVmjtHL7P9/z4B\n'
    'KdODlpb5Vx+54+Fa19vpgXEtHgfAgGW9DjlZMtl4wYTqyGAoa+SLuehjAQsxT8M1\n'
    'BXqfMJwE7D9XHjxkqCvd93UGgP+Yxe6H+HczJeA05dFLzC87qdM45R5c74k=\n'
    '-----END RSA PRIVATE KEY-----'
  ```
  Also, here are some examples if prefer to store your private key as a string
  but within an environment variable.
  ```sh
  # Local env example
  CF_PRIVATE_KEY="$(cat your-private-key.pem)"

  # Heroku env
  heroku config:set CF_PRIVATE_KEY="$(cat your-private-key.pem)"
  ```

## Examples
### Creating a signed URL
By default the URL will expire after half an hour.
```js
// ESM: import { getSignedUrl } from 'aws-cloudfront-sign'
const cf = require('aws-cloudfront-sign')
const options = {keypairId: 'APKAJM2FEVTI7BNPCY4A', privateKeyPath: '/foo/bar'}
const signedUrl = cf.getSignedUrl('http://xxxxxxx.cloudfront.net/path/to/s3/object', options);
console.log('Signed URL: ' + signedUrl);
```

### Creating signed cookies
```js
// ESM: import { getSignedCookies } from 'aws-cloudfront-sign'
const cf = require('aws-cloudfront-sign')
const options = {keypairId: 'APKAJM2FEVTI7BNPCY4A', privateKeyPath: '/foo/bar'}
const signedCookies = cf.getSignedCookies('http://xxxxxxx.cloudfront.net/*', options);

// You can now set cookies in your response header. For example:
for(var cookieId in signedCookies) {
 res.cookie(cookieId, signedCookies[cookieId]);
}
```

[moment_docs]: http://momentjs.com/docs
[cf_keypair_docs]: http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-trusted-signers.html#private-content-creating-cloudfront-key-pairs


CHANGELOG.md
[jasonsims]: https://github.com/jasonsims
[mpalmerlee]: https://github.com/mpalmerlee
[gregt590]: https://github.com/gregt590
[tbuchok]: https://github.com/tbuchok
[alexsaves]: https://github.com/alexsaves
[fponticelli]: https://github.com/fponticelli
[darrendao]: https://github.com/darrendao
[benburkhart1]: https://github.com/benburkhart1
[dnjohnson]: https://github.com/dnjohnson
[clarkie]: https://github.com/clarkie
[caub]: https://github.com/caub
[joelwass]: https://github.com/joelwass

## 3.0.2
* bugfix: fix empty package
## ~~3.0.1 (2023-07-12)~~ deprecated
* bugfix: remove leftover debug log [#73](https://github.com/jasonsims/aws-cloudfront-sign/issues/73)
## 3.0.0 (2023-07-10)
* feature: added support for TypeScript
* feature: added support for ES Modules
* development: engines updated to current LTS (v18)
* development: added commit lint
* development: migrated test suite to jest
* development: added funding URL
* development: removed lodash dependency
* development: added .nvmrc

## 2.2.0 (2016-11-09)
* development: [@caub][caub]: Simplified expiry and normalization logic
* development: [@joelwass][joelwass]: Added additional error handling for private key string

## 2.1.2 (2016-02-26)
* bugfix: [@clarkie][clarkie]: Fix undeclared variable
* development: [@jasonsims][jasonsims]: Update jshintrc to error on undeclared variables

## 2.1.1 (2016-02-24)
* feature: [@jasonsims][jasonsims]: Added support for NPM tonic examples

## 2.1.0 (2016-02-24)
* feature: [@benburkhart1][benburkhart1]: Added `ipRange` option to support source IP restrictions
* feature: [@darrendao][darrendao]: Added `getSignedCookies` API
* feature: [@dnjohnson][dnjohnson]: Updated the signed URL to also include the policy

## 2.0.1 (2015-04-21)
* develop: [@fponticelli][fponticelli]: Fix typo in usage examples
* develop: [@jasonsims][jasonsims]: Switch to TravisCI
* develop: [@jasonsims][jasonsims]: Add more descriptive usage examples

## 2.0.0 (2015-03-29)
* feature: [@jasonsims][jasonsims]: Add support for momentjs
* bugfix: [@jasonsims][jasonsims]: Fix malformed query param issue
* develop: [@jasonsims][jasonsims]: Misc API refactors
* develop: [@jasonsims][jasonsims]: Refactor of the test suite
* develop: [@jasonsims][jasonsims]: Add more description to API options in README.md

## 1.2.0 (2014-12-31)
* development: [@alexsaves][alexsaves]: Supporting JavaScript dates for expireTime
* development: [@alexsaves][alexsaves]: Added range and type checking
* development: [@alexsaves][alexsaves]: Added `getSignedRTMPUrl` for conveniently generating RTMP style urls

## 1.1.0 (2014-11-26)
* bugfix: [@gregt590][gregt590]: Normalize AWS signature strings
* development: [@tbuchok][tbuchok]: Add initial unit tests
* development: [@jasonsims][jasonsims]: Add tests for AWS signature
* development: [@jasonsims][jasonsims]: Cleanup code style

## 1.0.0 (2014-06-07)
* bugfix: [@mpalmerlee][mpalmerlee]: Fix issue with `expireTime` param
* development: [@mpalmerlee][mpalmerlee]: Make `gitSignedUrl` synchronous
* development: [@mpalmerlee][mpalmerlee]: Remove moment dependency


lib/cloudfrontUtil.js
/**
 * Cloudfront URL signature utility.
 */

var url = require('url');
var crypto = require('crypto');
var fs = require('fs');
var util = require('util');
var _ = require('lodash');
var CannedPolicy = require('./CannedPolicy');

/**
 * Build an AWS signed URL.
 *
 * @param {String} CloudFront resource URL
 * @param {Object} Signature parameters
 * @return {String} Signed CloudFront URL
 */
function getSignedUrl(cfUrl, params) {
  var privateKey = _getPrivateKey(params);
  var policy = _createPolicy(
    cfUrl, _getExpireTime(params), _getIpRange(params));
  var signature = _createPolicySignature(policy, privateKey);
  var policyStr = Buffer.from(policy.toJSON()).toString('base64');

  // Parse the cloudfront URL so we can add the querystring values required by
  // AWS signed URLs. We need to assign an empty string to the `search`
  // property so that the object value of `.query` is used when `.format()`
  // is called.
  var parsedUrl = url.parse(cfUrl, true);
  parsedUrl.search = '';
  _.extend(parsedUrl.query, {
    'Expires': policy.expireTime,
    'Policy': normalizeBase64(policyStr),
    'Signature': normalizeBase64(signature),
    'Key-Pair-Id': params.keypairId
  });

  // Return a formatted URL string with signature.
  return parsedUrl.format();
}

/**
 * Build an AWS signed RTMP URL assuming your distribution supports this.
 *
 * @param {String} Cloudfront domain
 * @param {String} S3 key
 * @param {Object} Signature parameters
 * @return {Object} Cloudfront server path and stream name with RTMP formatting
 */
function getSignedRTMPUrl(domainname, s3key, params) {

  if (!domainname || domainname.indexOf('/') > -1) {
    throw new Error(
      'Supplied domain name doesn\'t look right. ' +
      'Example: \'xxxxxxxx.cloudfront.net\'. Omit \'http\' and any paths.'
    );
  }

  if (!s3key || s3key.length === 0 || s3key.charAt(0) === '/') {
    throw new Error(
      'Supplied s3 key doesn\'t look right. ' +
      'Example: \'myfolder/bla.mp4\'. Omit preceding slashes or hostnames.'
    );
  }

  return {
    rtmpServerPath: 'rtmp://' + domainname + '/cfx/st',
    rtmpStreamName: module.exports.getSignedUrl(s3key, params)
  };
}

/**
 * Build list of cookies for the specified resource
 *
 * @param {String} CloudFront resource URL
 * @param {Object} Signature parameters
 * @return {String} List of CloudFront cookies
 */
function getSignedCookies(cfUrl, params){
  var privateKey = _getPrivateKey(params);
  var policy = _createPolicy(
    cfUrl, _getExpireTime(params), _getIpRange(params));
  var signature = _createPolicySignature(policy, privateKey);
  var policyStr = Buffer.from(policy.toJSON()).toString('base64');

  var cookies = {};
  cookies['CloudFront-Policy'] = normalizeBase64(policyStr);
  cookies['CloudFront-Signature'] = normalizeBase64(signature);
  cookies['CloudFront-Key-Pair-Id'] = params.keypairId;

  return cookies;
}

/**
 * Create a URL safe Base64 encoded string.
 *
 * This function will replace all characters that are invalid in a URL query
 * string with characters that are. AWS will translate these characters back to
 * their original value before processing.
 *
 * For more information, see
 * http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-creating-signed-url-canned-policy.html
 *
 * @param {String}: Base64 encoded string
 */
function normalizeBase64(str) {
  return str
    .replace(/\+/g, '-')
    .replace(/=/g, '_')
    .replace(/\//g, '~');
}

/**
 * Normalize all invalid AWS signature characters.
 *
 * For more information on AWS signatures, see
 * http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-creating-signed-url-canned-policy.html
 *
 * @deprecated

 */
function normalizeSignature(sig) {
  console.log(
    'Deprecation Warning: "normalizeSignature" will soon be removed. Please ' +
    'use "normalizeBase64"');
  return normalizeBase64(sig);
}

/**
 * Create an AWS policy.
 *
 * @private
 * @param {string} CloudFront resource URL
 * @param {Number|Moment|Date} URL expiration time
 * @param {string} Source IP range in CIDER notation
 * @returns {Object} AWS policy object
 */
function _createPolicy(cfUrl, expireTime, ipRange) {
  // If an expiration time isn't set, default to 30 minutes.
  var defaultExpireTime = Math.round(Date.now() + 1800000);
  expireTime = expireTime || defaultExpireTime;

  return new CannedPolicy(cfUrl, expireTime, ipRange);
}

/**
 * Create a policy signature.
 *
 * @private
 * @param {Object} AWS policy
 * @param {string} private key string for signature
 * @returns {string} Base64 encoded signature
 */
function _createPolicySignature(policy, privateKey) {
  var sign = crypto.createSign('RSA-SHA1');
  sign.update(policy.toJSON());

  return sign.sign(privateKey, 'base64');
}

/**
 * Return the expire time in milliseconds.
 *
 * This function will return the `expireTime` value in milliseconds. If the
 * value is not a number or a string it will use Object.valueOf
 *
 * @see http://momentjs.com/docs
 *
 * @param {Object} opts - Options object containing the expireTime
 * @return {Number} expireTime - Expire time represented in milliseconds
 */
function _getExpireTime(opts) {
  return +opts.expireTime || null;
}

/**
 * Helper function for retrieving the IP range from the params object.
 * @private
 */
function _getIpRange(opts) {
  return opts.ipRange || null;
}

/**
 * Helper function for retrieving the private key from the params object.
 * @private
 */
function _getPrivateKey(params) {
  var privateKeyString = params.privateKeyString;
  var pem;

  if (params.privateKeyPath) {
    pem = fs.readFileSync(params.privateKeyPath);
    privateKeyString = pem.toString('ascii');
  }

  var newLinePattern = /\r|\n/;
  var lineBreakExists = newLinePattern.test(privateKeyString);
  if (!lineBreakExists) {
      throw new Error('Invalid private key string, must include line breaks');
  }

  return privateKeyString;
}


exports.getSignedCookies = getSignedCookies;
exports.getSignedUrl = getSignedUrl;
exports.getSignedRTMPUrl = getSignedRTMPUrl;
exports.normalizeSignature = normalizeSignature;
exports.normalizeBase64 = normalizeBase64;
exports._getExpireTime = _getExpireTime;
exports._getIpRange = _getIpRange;
